File: src/domain/baseline-vectors.js
// ═══════════════════════════════════════════════════════════
// BaselineVectors — Painel de Vetores Configuráveis (B4)
// ═══════════════════════════════════════════════════════════
// Antes de rodar o baseline (5 melhores voltas pra evidência do Advisor),
// o engenheiro escolhe QUAIS dimensões considerar na seleção:
//   • pneu         { marca, composto, nivelDesgaste }
//   • ambiente     { tempPistaRange, tempArRange, clima }
//   • piloto       { apenasOMesmo, userId }
//   • carroConfig  { apenasOMesmo, carConfigId }
//   • dia          { apenasMesmoDia }
//
// Um preset fica salvo em meta (Dexie) sob key `baseline:preset:<trackId>`.

import { db } from '../core/db.js';

const KEY_PREFIX = 'baseline:preset:';

export const DEFAULT_VECTORS = Object.freeze({
  pneu:        { igualarMarca: true, igualarComposto: true, nivelMax: 2 },
  ambiente:    { tempPistaTol: 5, tempArTol: 5, apenasClima: true },
  piloto:      { apenasMesmoPiloto: true },
  carroConfig: { apenasMesmoCarroConfig: true },
  dia:         { apenasMesmoDia: false },
  minVoltas:   5,
});

export const BaselineVectors = {
  async getPreset(trackId) {
    const row = await db.meta.get(KEY_PREFIX + trackId);
    return row?.value || DEFAULT_VECTORS;
  },

  async setPreset(trackId, preset) {
    const value = { ...DEFAULT_VECTORS, ...preset };
    await db.meta.put({ key: KEY_PREFIX + trackId, value });
    return value;
  },
};

/**
 * Filtra candidatos de laps pra baseline a partir do preset.
 * @param {Array} candidateLaps - laps de várias sessions já enriquecidas com {session, env}
 * @param {Object} ref - { sessionId, env, userId, carConfigId, timestamp } do stint alvo
 * @param {Object} preset
 */
export function filterForBaseline(candidateLaps, ref, preset = DEFAULT_VECTORS) {
  const out = [];
  for (const c of candidateLaps) {
    if (!c.lap.valida) continue;
    const s = c.session;
    const e = c.env || {};
    const r = ref.env || {};

    if (preset.pneu.igualarMarca && r.composto && e.composto && r.composto !== e.composto) continue;
    if (preset.pneu.nivelMax != null && e.desgasteNivelD != null && e.desgasteNivelD > preset.pneu.nivelMax) continue;

    if (preset.ambiente.apenasClima && ref.clima && s.clima && ref.clima !== s.clima) continue;
    if (preset.ambiente.tempPistaTol != null && r.tempPista != null && e.tempPista != null) {
      if (Math.abs(r.tempPista - e.tempPista) > preset.ambiente.tempPistaTol) continue;
    }
    if (preset.ambiente.tempArTol != null && r.tempAr != null && e.tempAr != null) {
      if (Math.abs(r.tempAr - e.tempAr) > preset.ambiente.tempArTol) continue;
    }

    if (preset.piloto.apenasMesmoPiloto && ref.userId && s.userId !== ref.userId) continue;
    if (preset.carroConfig.apenasMesmoCarroConfig && ref.carConfigId && s.carConfigurationId !== ref.carConfigId) continue;
    if (preset.dia.apenasMesmoDia && ref.timestamp && s.dataInicio) {
      const mesmo = new Date(ref.timestamp).toDateString() === new Date(s.dataInicio).toDateString();
      if (!mesmo) continue;
    }

    out.push(c);
  }
  out.sort((a, b) => (a.lap.tempoMs || 9e9) - (b.lap.tempoMs || 9e9));
  return out.slice(0, Math.max(preset.minVoltas || 5, 10));
}

File: ios/p1fast-core/Sources/P1FastCore/BaselineVectors.swift
// ═══════════════════════════════════════════════════════════
// BaselineVectors — port Swift de src/domain/baseline-vectors.js
// ═══════════════════════════════════════════════════════════
// Painel de Vetores Configuráveis (B4) — antes de rodar baseline,
// engenheiro escolhe quais dimensões considerar na seleção das
// 5 melhores voltas.
//
// O Dexie (`getPreset/setPreset`) fica fora do port — persistência
// separada do core, conforme padrão Frente 3. Só a função pura
// `filterForBaseline` + estruturas/preset são portadas.

import Foundation

public enum BaselineVectors {

    public struct PneuPreset: Codable, Sendable {
        public var igualarMarca: Bool
        public var igualarComposto: Bool
        public var nivelMax: Int?
        public init(igualarMarca: Bool = true, igualarComposto: Bool = true, nivelMax: Int? = 2) {
            self.igualarMarca = igualarMarca
            self.igualarComposto = igualarComposto
            self.nivelMax = nivelMax
        }
    }

    public struct AmbientePreset: Codable, Sendable {
        public var tempPistaTol: Double?
        public var tempArTol: Double?
        public var apenasClima: Bool
        public init(tempPistaTol: Double? = 5, tempArTol: Double? = 5, apenasClima: Bool = true) {
            self.tempPistaTol = tempPistaTol
            self.tempArTol = tempArTol
            self.apenasClima = apenasClima
        }
    }

    public struct PilotoPreset: Codable, Sendable {
        public var apenasMesmoPiloto: Bool
        public init(apenasMesmoPiloto: Bool = true) { self.apenasMesmoPiloto = apenasMesmoPiloto }
    }

    public struct CarroConfigPreset: Codable, Sendable {
        public var apenasMesmoCarroConfig: Bool
        public init(apenasMesmoCarroConfig: Bool = true) { self.apenasMesmoCarroConfig = apenasMesmoCarroConfig }
    }

    public struct DiaPreset: Codable, Sendable {
        public var apenasMesmoDia: Bool
        public init(apenasMesmoDia: Bool = false) { self.apenasMesmoDia = apenasMesmoDia }
    }

    public struct Preset: Codable, Sendable {
        public var pneu: PneuPreset
        public var ambiente: AmbientePreset
        public var piloto: PilotoPreset
        public var carroConfig: CarroConfigPreset
        public var dia: DiaPreset
        public var minVoltas: Int

        public init(pneu: PneuPreset = .init(),
                    ambiente: AmbientePreset = .init(),
                    piloto: PilotoPreset = .init(),
                    carroConfig: CarroConfigPreset = .init(),
                    dia: DiaPreset = .init(),
                    minVoltas: Int = 5) {
            self.pneu = pneu
            self.ambiente = ambiente
            self.piloto = piloto
            self.carroConfig = carroConfig
            self.dia = dia
            self.minVoltas = minVoltas
        }
    }

    public static let defaultPreset = Preset()

    // ── tipos de entrada ──

    public struct EnvSnapshot: Sendable {
        public var composto: String?
        public var desgasteNivelD: Int?
        public var tempPista: Double?
        public var tempAr: Double?
        public init(composto: String? = nil, desgasteNivelD: Int? = nil,
                    tempPista: Double? = nil, tempAr: Double? = nil) {
            self.composto = composto
            self.desgasteNivelD = desgasteNivelD
            self.tempPista = tempPista
            self.tempAr = tempAr
        }
    }

    public struct LapInfo: Sendable {
        public let valida: Bool
        public let tempoMs: Double?
        public init(valida: Bool, tempoMs: Double? = nil) {
            self.valida = valida; self.tempoMs = tempoMs
        }
    }

    public struct SessionInfo: Sendable {
        public let userId: String?
        public let carConfigurationId: String?
        public let dataInicio: Date?
        public let clima: String?
        public init(userId: String? = nil, carConfigurationId: String? = nil,
                    dataInicio: Date? = nil, clima: String? = nil) {
            self.userId = userId
            self.carConfigurationId = carConfigurationId
            self.dataInicio = dataInicio
            self.clima = clima
        }
    }

    public struct Candidate: Sendable {
        public let lap: LapInfo
        public let session: SessionInfo
        public let env: EnvSnapshot?
        public init(lap: LapInfo, session: SessionInfo, env: EnvSnapshot? = nil) {
            self.lap = lap; self.session = session; self.env = env
        }
    }

    public struct Reference: Sendable {
        public let userId: String?
        public let carConfigId: String?
        public let timestamp: Date?
        public let clima: String?
        public let env: EnvSnapshot?
        public init(userId: String? = nil, carConfigId: String? = nil,
                    timestamp: Date? = nil, clima: String? = nil,
                    env: EnvSnapshot? = nil) {
            self.userId = userId; self.carConfigId = carConfigId
            self.timestamp = timestamp; self.clima = clima; self.env = env
        }
    }

    /// Filtra candidatos a baseline a partir do preset.
    /// Espelho 1:1 da função JS — ordena por tempoMs ascendente
    /// e devolve até `max(preset.minVoltas, 10)` candidatos.
    public static func filterForBaseline(candidates: [Candidate],
                                         ref: Reference,
                                         preset: Preset = defaultPreset) -> [Candidate] {
        var out: [Candidate] = []
        for c in candidates {
            if !c.lap.valida { continue }
            let e = c.env ?? EnvSnapshot()
            let r = ref.env ?? EnvSnapshot()
            let s = c.session

            // Pneu
            if preset.pneu.igualarMarca, let rc = r.composto, let ec = e.composto, rc != ec { continue }
            if let nMax = preset.pneu.nivelMax, let dn = e.desgasteNivelD, dn > nMax { continue }

            // Ambiente — clima
            if preset.ambiente.apenasClima, let refClima = ref.clima, let sClima = s.clima, refClima != sClima { continue }
            // Ambiente — temperaturas
            if let tol = preset.ambiente.tempPistaTol, let rt = r.tempPista, let et = e.tempPista,
               abs(rt - et) > tol { continue }
            if let tol = preset.ambiente.tempArTol, let ra = r.tempAr, let ea = e.tempAr,
               abs(ra - ea) > tol { continue }

            // Piloto
            if preset.piloto.apenasMesmoPiloto, let refUser = ref.userId, s.userId != refUser { continue }
            // Carro/configuração
            if preset.carroConfig.apenasMesmoCarroConfig, let refCfg = ref.carConfigId,
               s.carConfigurationId != refCfg { continue }
            // Dia
            if preset.dia.apenasMesmoDia, let refTs = ref.timestamp, let sTs = s.dataInicio {
                let cal = Calendar(identifier: .gregorian)
                let a = cal.startOfDay(for: refTs)
                let b = cal.startOfDay(for: sTs)
                if a != b { continue }
            }

            out.append(c)
        }

        out.sort { (a: Candidate, b: Candidate) in
            (a.lap.tempoMs ?? 9e9) < (b.lap.tempoMs ?? 9e9)
        }
        let limite = max(preset.minVoltas, 10)
        return Array(out.prefix(limite))
    }
}
